### ISSUE TITLE
Identity operator against constructor fails to narrow the type

### ISSUE THREAD

**Issue by gh-andre** [microsoft/pyright#9669]
**Describe the bug**

Trying to narrow a generic type via the identity operator fails in that all other alternatives are considered and errors are reported..

**Code or Screenshots**

```python
from typing import TypeVar, Generic
from datetime import datetime

class XW:
    def __init__(self, i: int) -> None:
        self.x: int = i
    
class YW:
    def __init__(self, s: str) -> None:
        self.s: str = s

class ZW:
    def __init__(self, d: datetime) -> None:
        self.d: datetime = d

T = TypeVar("T", XW, YW, ZW)

class C(Generic[T]):
    def __init__(self) -> None:
        self.l: list[T] = list[T]()

def fn(c: C[T], ctor: type[T]) -> C[T]:
    t: T

    if ctor is XW:
        t = ctor(123)
    elif ctor is YW:
        t = ctor("ABC")
    elif ctor is ZW:
        t = ctor(datetime.now()) 
    else:
        raise ValueError()
        
    c.l.append(t)

    return c

fn(C[XW](), XW)
fn(C[YW](), YW)
fn(C[ZW](), ZW)
```

For each of the constructor calls above these errors are reported:

> Argument of type "Literal[123]" cannot be assigned to parameter "s" of type "str" in function "__init__"
  "Literal[123]" is not assignable to "str"  (reportArgumentType)
Argument of type "Literal[123]" cannot be assigned to parameter "d" of type "datetime" in function "__init__"
  "Literal[123]" is not assignable to "datetime"  (reportArgumentType)

> Argument of type "Literal['ABC']" cannot be assigned to parameter "i" of type "int" in function "__init__"
  "Literal['ABC']" is not assignable to "int"  (reportArgumentType)
Argument of type "Literal['ABC']" cannot be assigned to parameter "d" of type "datetime" in function "__init__"
  "Literal['ABC']" is not assignable to "datetime"  (reportArgumentType)

> Argument of type "datetime" cannot be assigned to parameter "i" of type "int" in function "__init__"
  "datetime" is not assignable to "int"  (reportArgumentType)
Argument of type "datetime" cannot be assigned to parameter "s" of type "str" in function "__init__"
  "datetime" is not assignable to "str"  (reportArgumentType)

Strangely enough, adding this line before the `if` statement makes all errors go away. 

```python
    if issubclass(ctor, float):
        raise ValueError()
```

I don't see any additional objects in the inheritance chain - printing `__mro__` of one of those constructors outputs the expected list:
```
(<class '__main__.XW'>, <class 'object'>)
```

**VS Code extension or command-line**

I ran into the original error in VS2022, with Pylance 17.0.24269.5. VS doesn't have much in the settings for type checking, except that it's set up as _strict_. The example above shows errors in Pyright playground with the latest version of Pyright and all versions of Python, but the version I'm using is Python 3.10.

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug.

Using the `is` operator isn't safe to narrow in this case.

You're using a value-constrained type variable here. I recommend against using these. They're poorly-specified in the typing spec. Associated behaviors will differ between type checkers, and it's not clear to me that we'll ever be able to full specify them in a way that is consistent with other aspects of the type system. Avoid them if possible. Use unions and overloads where possible.

A value-constrained type variable does not constrain a value to one of the specified constraints. It constrains it to one of the constrains *or a subtype thereof*. That means a subclass of `XW` in your example would satisfy the constraint of `XW`. It's also possible in your example for a type to be a subclass of both `XW` and `YW`, in which case the intended behavior of a value-constrained type variable is unclear (see my point above).

To make your code robust and type safe, you should use `issubclass` rather than `is`. If you make this change, it will type check without errors:

```python
def fn(c: C[T], ctor: type[T]) -> C[T]:
    t: T

    if issubclass(ctor, XW):
        t = ctor(123)
    elif issubclass(ctor, YW):
        t = ctor("ABC")
    elif issubclass(ctor, ZW):
        t = ctor(datetime.now())
    else:
        raise ValueError()

    c.l.append(t)

    return c
```

**Comment by gh-andre** [comment]
Thanks for responding. 

`issubclass` cannot be applied against `TypedDict` instances, where is what my main code needs to use. 

With regards to constraints, Python describes it as specific types, which is contrasted further in their examples against inherited types:

> Using a constrained type variable, however, means that the TypeVar can only ever be solved as being exactly one of the constraints given

So, one may not like using them, but if you want to pick a class out of a hierarchy or use `TypedDict` instances, then static type checkers should resolve identity as described within the specific branch of the conditional. So, your point is valid for the entire conditional statement - the final `else` may contain some of the derived classes, but within each branch it should be considered just as that class from the identity operator.

**Comment by gh-andre** [comment]
Also, that rogue `issubclass()` against `float` is also showing that something is going on there with types. I wouldn't expect it to have any effect if everything was working as inttended.

**Comment by erictraut** [comment]
Thanks for the additional details about your use case.

As I said above, I discourage people from using value-constrained type variables. There's almost always a better approach. In your case, you could use a union combined with a 
I think it would be theoretically safe to apply type narrowing in this case. It's something I might consider adding in the future if there's sufficient demand from pyright users. 

I'll point out that mypy also doesn't narrow in this case.

**Comment by gh-andre** [comment]
> In your case, you could use a union combined with an overload.

A union creates a list of A or B and I need a list of A or a list of B, which can only be achieved via constraints. In my case I have control over what is listed in constraints, so it's safe for me because I can catch dev mismatches in terms of inherited classes via `else` and with a good unit test coverage I'm covered fairly well in terms of unexpected behavior.

> I'll point out that mypy also doesn't narrow in this case.

MyPy is completely broken against constraints. You can see my bug on the topic in their issues - they don't even narrow simple types like `int`, `str`.

> I might consider adding in the future if there's sufficient demand from pyright users.

I understand. Thank you for considering. I appreciate it.

FWIW, the hack with `issubclass(ctor, float)` started working in Pyright v1.1.389 - I can see it reporting errors in 1.1.388 for this code, so something has changed in that version that altered this behavior.

**Comment by erictraut** [comment]
The design you're using here definitely constrains your options. It's an unusual design choice, which is probably why you're feeling the need to use approaches that are not well specified or supported. You might want to consider alternative designs.
